### PR TITLE
Provide an API in RequestLog to obtain traceId and spanId.

### LINKED ISSUE #1824: Provide a way to access Zipkin span ID via `RequestLog`.
Currently, Zipkin's `TraceContext` is stored as an attribute in `RequestContext`:

- https://github.com/line/armeria/blob/armeria-0.86.0/zipkin/src/main/java/com/linecorp/armeria/common/tracing/RequestContextCurrentTraceContext.java#L69-L70

We could provide some utility methods for retrieving some useful information from it, such as `traceId`, `spanId` and `parentId`.

### PR DESCRIPTION
**Pull request by KonaEspresso94**
Motivation:

#1824 

Currently, `RequestLog` does not expose information such as `traceId` and `spanId`.
Consequently, when `traceId` and `spanId` are needed in `AccessLogWriter`, we have to be directly retrieved via `com.linecorp.armeria.internal.common.brave.TraceContextUtil.traceContext`.

Introducing an API in RequestLog to access trace information would be beneficial.

Modifications:

- Add `TracingContext` as a holder class for traceContext log information.
- For `tracingContext`, a getter is added to `RequestOnlyLog` and a setter to `RequestLogBuilder`.
- In `BraveService::serve` and `BraveClient::execute`, the method `ctx.logBuilder().tracingContext()` is invoked before proceeding to the delegate to populate the tracing information.

Result:

- Closes #1824

### CODE REVIEW

File: core/src/main/java/com/linecorp/armeria/common/logging/DefaultRequestLog.java [hunk 1]
```diff
@@ -598,6 +601,17 @@ public void endResponseWithLastChild() {
...
     }
 
+    @Nullable
+    @Override
+    public TracingContext tracingContext() {
+        return this.tracingContext;
+    }
+
+    @Override
+    public void tracingContext(TracingContext tracingContext) {
```
Comment (minwoox) [inline core/src/main/java/com/linecorp/armeria/common/logging/DefaultRequestLog.java:611]: I'm uncertain about adding the `TracingContext` to the core module because different tracing libraries may have different fields and meanings. For instance, in Pinpoint, the traceId represents a collection of keys (as outlined here: https://d2.naver.com/helloworld/1194202).

Instead, what do you think about simply providing a utility class in the brave module to facilitate easy extraction of properties, rather than adding TracingContext to the core module? We can consider adding the API to the core module later, once it's well-stabilized.

**Review by minwoox — COMMENTED** [review]
Hi! Thanks a lot for the PR. I left my opinion. 😉 
Could you also add some tests, please?

File: brave/src/main/java/com/linecorp/armeria/internal/common/brave/SpanConverter.java [hunk 2]
```diff
@@ -0,0 +1,37 @@
...
+
+import brave.Span;
+import brave.propagation.TraceContext;
+
+public final class SpanConverter {
+
+    public static TracingContext toLogTracingContext(Span span) {
+        final TraceContext traceContext = span.context();
+
+        return TracingContext.of(traceContext.traceIdHigh(), traceContext.traceId(),
```
Comment (codefromthecrypt) [inline brave/src/main/java/com/linecorp/armeria/internal/common/brave/SpanConverter.java:29]: should add a benchmark if this will be used in a hot path, to make sure any overhead is fine. IIRC brave caches the string versions of IDs, but if nothing uses these data it would be useless overhead

Comment (minwoox) [reply brave/src/main/java/com/linecorp/armeria/internal/common/brave/SpanConverter.java:29]: That's a good idea. 👍 
@codefromthecrypt How do you feel about this comment because you are the expert on distributed tracing? https://github.com/line/armeria/pull/5414#discussion_r1488938475
Can all tracing frameworks and libraries share the `TracingContext`?

Comment (codefromthecrypt) [reply brave/src/main/java/com/linecorp/armeria/internal/common/brave/SpanConverter.java:29]: there are differences between libraries. For example, the localRoot is not a common concept, it is only available in brave (unless someone copied it later). also some don't propagate parent ID. I think trace and span ID are pretty common though.

Comment (codefromthecrypt) [reply brave/src/main/java/com/linecorp/armeria/internal/common/brave/SpanConverter.java:29]: brave has built-in context stuff, and maybe otel does too? not sure. It might be better to configure this stuff per library vs make an abstraction https://github.com/openzipkin/brave/tree/master/context/slf4j

Comment (minwoox) [reply brave/src/main/java/com/linecorp/armeria/internal/common/brave/SpanConverter.java:29]: Thanks @codefromthecrypt for the suggestion. 😉 
Then, I think we'd better put this in the brave module and abstract it later. @KonaEspresso94 Could you do that, please?

Comment (codefromthecrypt) [reply brave/src/main/java/com/linecorp/armeria/internal/common/brave/SpanConverter.java:29]: yep, actually the context integration was some of the trickiest code, and thankfully had a good design bud @anuraaga at the time

Comment (minwoox) [reply brave/src/main/java/com/linecorp/armeria/internal/common/brave/SpanConverter.java:29]: @KonaEspresso94 Please forget about my previous comments. 😓 
Let's add `TracingContext` to the core module as you did.
Before proceeding, as suggested by @codefromthecrypt, it's worth noting that not all properties in the TracingContext are common to other tracing frameworks and libraries. Could you conduct research on other platforms such as [OpenTelemetry](https://opentelemetry.io/) and [Micrometer tracing](https://docs.micrometer.io/tracing/reference/) to identify common properties?

**Comment by codefromthecrypt** [comment]
the reason things like this are tricky is because there can be subspans in use, and in that case unless you do something like I mentioned, the span IDs will be wrong (for example, you start a span inside a server handler). So, make sure that where this static-copy of trace/span IDs is in a place that it could only ever be a single span. This type of thing is a lot tricker than it seems, especially when things are deferred to threads. We already have context integration in, and that can get the implicit span ID.

TL;DR;

if you must do somethning like this.. make integration tests that will fail if there is a subspan for each one, and that this test runs both in brave and micrometer, and that the overhead is tolerable for all people who use this. cc @kojilin

**Comment by codefromthecrypt** [comment]
As the issue referenced was from 2019 I don't think we need to rush this. Please focus on the integration tests and not arbitrary compat with libraries not even integrated. This particular thing has a potential to mislead, which is the worst thing to do in logs. Let's make sure it is done right, with tests, or deferred to built-in logic.

File: core/src/main/java/com/linecorp/armeria/common/logging/TracingContext.java [hunk 3]
```diff
@@ -0,0 +1,150 @@
...
+
+import com.google.common.base.MoreObjects;
+
+import com.linecorp.armeria.common.Request;
+import com.linecorp.armeria.common.annotation.Nullable;
+
+/**
+ * A holder class which has the tracing(e.g. zipkin) information.
+ */
+public final class TracingContext {
```
Comment (codefromthecrypt) [inline core/src/main/java/com/linecorp/armeria/common/logging/TracingContext.java:29]: so here's my TL;DR; info. If this happens, this should be an interface, which will allow lazy loading of details only when present, and also a no-op impl. don't worry about random properties, and focus on traceID/spanId and doing it here vs alternates like via MDC (status quo for most)

Here are the edge cases to consider:

1. there is no trace at all, so should TracingContext be returned as null in this case? or nullable traceId and spanId
2. documentation about this type about which span it is about. For example, is this the server request? or any request?
3. use cases this supports, exactly. For example, is this for setting MDC? If so, when would you use this vs normal MDC integration
4. how to disable this, and what is the fallback (e.g maybe behavior of unsampled is same as disabled)

Tests should cover these questions, and a benchmark to show we aren't accidentally copying data per request when there is no trace or this feature is disabled.

**Review by codefromthecrypt — COMMENTED** [review]
hope this clarifies my myriad of things. As I know everyone here are performance and quality focused, thinking through these, you can steer towards a path forward!

**Comment by minwoox** [comment]
> the reason things like this are tricky is because there can be subspans in use,

I completely overlooked that aspect. I had assumed a one-to-one relationship between a span and an Armeria RequestContext. Could you provide a specific example where multiple spans can be created for a RequestContext?

> if you must do somethning like this.. make integration tests that will fail if there is a subspan for each one, and that this test runs both in brave and micrometer, and that the overhead is tolerable for all people who use this.

I believe this was included for convenience, meaning that it's not mandatory for us to provide it. If you feel that it adds unnecessary complexity without any clear advantage, we may consider halting the implementation of this feature. cc @line/dx

**Comment by codefromthecrypt** [comment]
> Could you provide a specific example where multiple spans can be created for a RequestContext?

Any outgoing request, while in flight, would be a different span ID (most commonly the client or message producer)

https://github.com/line/armeria/blob/7d2ab6487bec29bf8aae041da075ddf6f7199238/site/src/pages/docs/advanced-zipkin.mdx#L21

Also, any time someone uses the tracer to start an ad-hoc span.

p.s. main reason I ask about the concrete use cases (what makes this convenient) is I might be able to help offer a way to accomplish the same with no new API. Right now, I just really don't know what this is intended to be used to accomplish, so hard to advise.

If it is only about `AccessLogWriter`, seems we can drive this easier (like the other context utilities in brave.. they help with MDC, and even flight recorder in the same way). So, mainly trying to narrow if this is about `AccessLogWriter` or something else.  If `AccessLogWriter`, then we can evaluate the normal context integration which is customizable to choose which fields even baggage properties.

**Comment by codefromthecrypt** [comment]
also, if it is about AccessLogWriter, then the span ID should be exactly that of the server span, so it removes the whole insecurity about what the span ID should be, when.

File: brave/src/main/java/com/linecorp/armeria/client/brave/BraveClient.java [hunk 4]
```diff
@@ -127,6 +128,8 @@ public HttpResponse execute(ClientRequestContext ctx, HttpRequest req) throws Ex
             ctxExtension.hook(() -> currentTraceContext.newScope(span.context()));
         }
 
+        ctx.logBuilder().tracingContext(SpanConverter.toLogTracingContext(span));
```
Comment (codefromthecrypt) [inline brave/src/main/java/com/linecorp/armeria/client/brave/BraveClient.java:131]: @minwoox @KonaEspresso94 if this is only for access logs, my main question here is that is this a push/pop sort of thing. so for example, is this masking the server request log? or is it a different object?

Note: the code above this pushes the span context already, just to integrate with brave instead of the log builder.
```

        final RequestContextExtension ctxExtension = ctx.as(RequestContextExtension.class);
        if (currentTraceContext != null && !span.isNoop() && ctxExtension != null) {
            // Make the span the current span and run scope decorators when the ctx is pushed.
            ctxExtension.hook(() -> currentTraceContext.newScope(span.context()));
        }
```

where you see the "hook" part seems like the place to do something like this, as it only triggers where there is an active span (not no-op due to disabled or otherwise). So if we go with this change, I think we do it inside the same if.

second part: push something derived from the span context not the span, as the context is ultimately what has the properties needed.

third part: If we want to push the context for retrieval of IDs, basically sounds fine, just we need to document what value someone should expect and have unit tests for client spans and server spans (that the IDs are what we think they should be). The documentation should say that the TraceContext (really should be that not TracingContext but that's a nit) is only present when there is a sampled span.

If those IDs are the "current client or server span" then copying off the values works fine, because this logic inside brave it will be cached, so not create additional garbage.

```java
  volatile String spanIdString; // Lazily initialized and cached.

  /** Returns the hex representation of the span's ID */
  public String spanIdString() {
    String r = spanIdString;
    if (r == null) {
      r = spanIdString = toLowerHex(spanId);
    }
    return r;
  }
```

So, here's the summary advice for implementing this for AccessLog:

1. decide whether `TracingContext (adrian likes TraceContext better) ` should be a class or an interface. probably if we want to guarantee the value doesn't change, class is best despite the cost of populating it.
2. document the `TracingContext` holds the last client or server span ID, and is only present when when there is a sampled span (so null when that's not the case for any reason)
3. move the populating part to inside the same if statement as the ctxExtension.hook, so no overhead unless there's a real context
4. add integration tests for client and server together.
5. showcase how this is actually used (most important part) as there's already more flexible log integration with SLFJ and brave.

**Comment by codefromthecrypt** [comment]
last thing to think through is that this discussion is easy when it is only about traceID and spanId, but in brave, we started with traceID and spanID, and later people ask for baggage properties (user-id properties etc.) we get into things that are not cached. So, the reason it is *easy* as 1-5 is because we are stopping at traceID/spanID.. if we go further, eventually we will recreate brave's context integration, except not have the advantages brave has due to internal tricks. So, think about if the real use case for this will stop at span IDs.

Here's a concrete example. https://github.com/openzipkin/brave-example/tree/master/armeria

The armeria brave example currently has this:

```
      <pattern>%d{HH:mm:ss.SSS} [%thread] [%X{userName}] [%X{traceId}/%X{spanId}] %-5level %logger{36} - %msg%n</pattern>
```

The `userName` property is custom case format from the inbound property.. this sort of problem is common case in log correlation.

This is simply configured like this:

```
  /** Allows log patterns to use {@code %{traceId}} {@code %{spanId}} and {@code %{userName}} */
  static ScopeDecorator correlationScopeDecorator() {
    return MDCScopeDecorator.newBuilder()
        .add(CorrelationScopeConfig.SingleCorrelationField.create(USER_NAME)).build();
  }
```

And it works without any of this today. This is why I suggest looking hard about acceslogger and see what the end goal is and work backwards from that to find the solution that is the least new code to maintain and makes least issues where something that starts simple ends up accumulating a lot of stuff, as the simple thing teases more and more features.

**Comment by KonaEspresso94** [comment]
@minwoox @codefromthecrypt 

Hello, I apologize for the delayed response to your comments.

Yes, right. The purpose of this PR was indeed to make it easier to access traceId and spanId at the boundaries of internal and external interactions when logging AccessLog. 
I overlooked that the `Log` object encompasses not just the information occurring at the boundaries of Server or Client's internal and external interactions but also includes contexts for retries and subroutines used during the request processing. In cases where `Log` is accessed for other purposes, it might seem necessary to obtain all trace information related to that request within the application, not just the spans or trace information occurring at the internal/external boundaries.

Also, I agree that the boundary between what is considered a tracing context and what is not is unclear, it becomes influenced by various frameworks. If constructed improperly from the beginning, there's a risk it could lead towards the need for reimplementation of trace functionalities in the future. Also, attempting to abstract access might introduce performance issues, which seems risky. Accessing through MDC, as mentioned by @codefromthecrypt, or the TraceContextUtil I noted in the PR, appears to be the most appropriate approach.

However, my motivation for undertaking this task was that I am not using MDC in my environment, and I had to sift through the armeria source code to find out about TraceContextUtil. Instead of proceeding with this task, suggesting the addition of an example using TraceContextUtil for the access log writer in the armeria examples might have been a better approach.

Thank you for the review. It has provided me good chance to consider much.
I will close this PR.

**Comment by codefromthecrypt** [comment]
@KonaEspresso94 thanks also for your consideration. Whenever I do anything new in tracing with brave, I use https://github.com/openzipkin/brave-example/tree/master/armeria, so if you have a chance you can add access logging to it. Possibly, this will help save your experience as well? Once that's in, maybe I can help update docs here on the same.

**Comment by minwoox** [comment]
Thanks @codefromthecrypt for all your valuable advice! 😊
I believe the end goal was likely to create a tracing data class that encompasses properties commonly utilized by popular tracing frameworks.
However, as you've pointed out, there are some challenges with this approach. Let's table this discussion for now and revisit it later if the need arises.

Thank you, @KonaEspresso94, for your efforts on this. Please also take into account the insights shared by @codefromthecrypt.

**Comment by codefromthecrypt** [comment]
ps if this gets resurrected, I think basically we need to add keys `traceId` `spanId` to this and show it works in a custom format https://armeria.dev/docs/server-access-log#retrieving-values-from-requestlog

**Comment by codefromthecrypt** [comment]
finally as we spent a lot of time on this topic, I added the access logger integration myself to the example. It actually doesn't work out of the box, as I thought that it would have. Maybe some opinions here can help https://github.com/openzipkin/brave-example/pull/115
